File: api/supplier_route.py
from logger import configure_logger
from api.api_utils import format_response
from flask import Blueprint, request, url_for
from suppliers.suppliers import Supplier


api_logger = configure_logger(__name__, True)
supplier_routes = Blueprint("supplier_routes", __name__)

# CREATE NEW SUPPLIER
@supplier_routes.route("/v1/supplier", methods=["POST"])
def add_supplier():
    data = request.get_json()
    required_fields = ["supplier_name", "supplier_address", "supplier_contact"]
    
    missing_fields = [field for field in required_fields if field not in data]
    if missing_fields:
        api_logger.error(f"Missing required fields: {','.join(missing_fields)}")
        return format_response(400, f"Missing required fields: {','.join(missing_fields)}")

    new_supplier = Supplier(data["supplier_name"], data["supplier_address"], data["supplier_contact"])
    try:
        new_supplier.save_to_db()
        response = format_response(201, "Supplier added successfully", new_supplier.to_dict())
        response.headers["location"] = url_for("supplier_routes.get_supplier", supplier_id=new_supplier.id, _external=True)
        return response
    except Exception as e:
        api_logger.error(f"Failed to add supplier: {e}")
        return format_response(400, str(e))




@supplier_routes.route("/v1/supplier/<int:supplier_id>", methods=["GET"])
def get_supplier(supplier_id):
    supplier = Supplier.find_by_id(supplier_id)
    if supplier is None:
        return format_response(400, "Supplier Not Found")
    return format_response(200, "Supplier retrieved successfully", supplier.to_dict())





# GET INFORMATION
@supplier_routes.route("/v1/supplier/name/<string:name>", methods=["GET"])
def get_supplier_by_name(name):
    try:
        suppliers = Supplier.get_supplier_by_name(name)
        suppliers_dict = [supplier.to_dict() for supplier in suppliers]
        return format_response(200, "Supplier retrieved successfully", suppliers_dict)
    except Exception as e:
        api_logger.error(f"Failed to retrieve supplier: {e}")
        return format_response(500, str(e))

@supplier_routes.route("/v1/supplier/all", methods=["GET"])
def get_all_supplier():
    try:
        suppliers = Supplier.get_all_suppliers()
        suppliers_dict = [supplier.to_dict() for supplier in suppliers]
        return format_response(200, "Suppliers retrieved successfully", suppliers_dict)
    except Exception as e:
        api_logger.error(f"Failed to retrieve suppliers: {e}")
        return format_response(500, str(e))





# UPDATING
@supplier_routes.route("/v1/supplier/<int:id>/<string:field>", methods=["PUT"])
def update_supplier_field(id, field):
    new_value = request.json.get("new_value")
    supplier = Supplier.find_by_id(id)
    if supplier:
        if field in ["name", "address", "contact"]:
            supplier.update_field(field, new_value)
            return format_response(200, f"Supplier {field} updated successfully", supplier.to_dict())
        else:
            return format_response(400, f"Invalid field: {field}")
    else:
        return format_response(400, "Supplier not found")

File: archive/js/supplier_route.js
function appendData(row, supplier) {
	var order = ["id", "supplier_name", "supplier_address", "supplier_contact"];
	order.forEach(function (key) {
		row.append("<td>" + supplier[key] + "</td>");
	});
}

function sendRequest(url, type, data) {
	$.ajax({
		url: url,
		type: type,
		contentType: "application/json",
		data: JSON.stringify(data),
		success: function (response) {
			if (type == "DELETE") {
				$("#result").html(response.message);
			} else {
				var suppliers = response.data;
				var table = $("<table class='table'>");
				table.append(
					"<tr><th>ID</th><th>Name</th><th>Address</th><th>Contact</th></tr>"
				);
				suppliers.forEach(function (supplier) {
					var row = $("<tr>");
					appendData(row, supplier);
					table.append(row);
				});
				$("#result").html(table);
			}
		},
		error: function (response) {
			$("#result").html(JSON.stringify(response));
		},
	});
}

$("#delete-form").submit(function (event) {
	event.preventDefault();

	var id = $("#delete-id").val();

	sendRequest(`http://localhost:5000/v1/supplier/delete/${id}`, "DELETE", {});
});

$("#add-form").submit(function (event) {
	event.preventDefault();

	var name = $("#name").val();
	var address = $("#address").val();
	var contact = $("#contact").val();

	sendRequest(`http://localhost:5000/v1/supplier`, "POST", {
		supplier_name: name,
		supplier_address: address,
		supplier_contact: contact,
	});
});

$("#get-form").submit(function (event) {
	event.preventDefault();

	var id = $("#get-id").val();

	sendRequest(`http://localhost:5000/v1/supplier/${id}`, "GET", {});
});

$("#update-form").submit(function (event) {
	event.preventDefault();

	var id = $("#update-id").val();
	var field = $("#field").val();
	var newValue = $("#new-value").val();

	sendRequest(`http://localhost:5000/v1/supplier/${id}/${field}`, "PUT", {
		new_value: newValue,
	});
});

$("#get-all").click(function () {
	sendRequest(`http://localhost:5000/v1/supplier/all`, "GET", {});
});

$("#get-name-form").submit(function (event) {
	event.preventDefault();
	var name = $("#search-name").val();
	sendRequest(`http://localhost:5000/v1/supplier/name/${name}`, "GET");
});
